### PR TITLE
Improve thread safety of resource loading

### PR DESCRIPTION
**Pull request by RandomShaper**
During resource loading, including threaded loading, certain resources want to emit a `changed` signal. Since most of the `Object` API, including signals, is not thread-safe, there's the risk of this causing issues. I've actually seen it happening, but I can't be sure if not getting them anymore is due to this fix.

This PR will defer any emissions of that kind involved in loading that happen on a non-main thread. **UPDATE:** Bear in mind that during threaded loading, deferred calls are even more deferred, queued in a thread-specific call queue, until the resource is done loading, then flushed on the main thread.

**UPDATE 2:** An alternative would be letting `Object` itself defer signal-related calls during loads. However, I'm a bit afraid of adding a check there is too heavy for what so far seem to be few cases (more to discover, likely). Additionally, maybe at some point `Object` has to be made entirely thread-safe due to popular demand or something, and patches like this are not needed anymore.

### CODE REVIEW

**Comment by KoBeWi** [comment]
Shader changes cause values to not load properly at runtime:

https://github.com/godotengine/godot/assets/2223172/b6492965-af5b-4789-9514-d2c83f7bfa34

[BrokenShader.tscn.txt](https://github.com/godotengine/godot/files/14426183/BrokenShader.tscn.txt)

**Comment by RandomShaper** [comment]
@KoBeWi, thanks for reviewing. Can you confirm you have tested with the latest push I made to this branch? I'm asking because I realized issues and pushed fixes.

If you are already testing the latest, please share the MRP. I need it because with the latest changes I'm not seeing issues anymore on my side.

**Comment by KoBeWi** [comment]
It was with latest changes and I already shared MRP.

**Comment by RandomShaper** [comment]
> It was with latest changes and I already shared MRP.

I've been like reliable research, that is, doubly-bind:
1. I couldn't stop the MRP.
2. When I re-pushed to this branch, I did it from a source branch that hadn't been updated with the fixes.

It should be OK now.

File: core/io/resource.cpp [hunk 1]
```diff
@@ -152,12 +157,22 @@ bool Resource::editor_can_reload_from_file() {
 }
 
 void Resource::connect_changed(const Callable &p_callable, uint32_t p_flags) {
+	if (ResourceLoader::is_within_load() && !Thread::is_main_thread()) {
+		// Let the check and connection happen on the main thread, later, since signals are not thread-safe.
+		callable_mp(this, &Resource::connect_changed).bind(p_callable, p_flags).call_deferred();
```
Comment (KoBeWi) [inline core/io/resource.cpp:162]:
```suggestion
		callable_mp(this, &Resource::connect_changed).call_deferred(p_callable, p_flags);
```

File: core/io/resource.cpp [hunk 2]
```diff
@@ -152,12 +157,22 @@ bool Resource::editor_can_reload_from_file() {
...
+	}
 	if (!is_connected(CoreStringNames::get_singleton()->changed, p_callable) || p_flags & CONNECT_REFERENCE_COUNTED) {
 		connect(CoreStringNames::get_singleton()->changed, p_callable, p_flags);
 	}
 }
 
 void Resource::disconnect_changed(const Callable &p_callable) {
+	if (ResourceLoader::is_within_load() && !Thread::is_main_thread()) {
+		// Let the check and disconnection happen on the main thread, later, since signals are not thread-safe.
+		callable_mp(this, &Resource::disconnect_changed).bind(p_callable).call_deferred();
```
Comment (KoBeWi) [inline core/io/resource.cpp:173]:
```suggestion
		callable_mp(this, &Resource::disconnect_changed).call_deferred(p_callable);
```

**Review by KoBeWi — APPROVED** [review]
I did some testing and it doesn't seem to break anything now, but I'm not sure how to test whether it fixes something.

**Comment by RandomShaper** [comment]
The way I realized the issue was using multi-threaded loading, with `ShaderMaterial`s involved. On many runs I was getting different issues, the most frequent being a failed dev assertion in `Object::emit_signal()` or different kinds of crashes.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**